optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/cloner/DeepCloningUtils.java

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.domain.solution.cloner;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.ConcurrentMap;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.optaplanner.core.api.domain.solution.cloner.DeepPlanningClone;
+import org.optaplanner.core.impl.domain.common.ReflectionHelper;
+import org.optaplanner.core.impl.domain.solution.descriptor.SolutionDescriptor;
+
+public final class DeepCloningUtils {
+    // owningClass != declaringClass
+    // owningClass refers to the actual instance class, which may be a subclass of the declaring class
```

Comment:
Instead of having all static methods, can this class have a non-static field for the Memoizations maps and have all non-static methods? Is that reasonable for the clients of this class, to remember the class instance (instead of those Memoizations maps) until they're done with it (so done with the Memoizations) at which point it gets garbage collected.